### DISCUSSION TITLE
Strange driversigning with untrusted test-certificate and certificate that signed the RustDeskIddDriver.dll is also not the same certificate

### DISCUSSION THREAD

**Discussion by vitusb** [rustdesk/rustdesk#6444] [Q&A] [answered]
### Bug Description

Hello ...

I've executed the Windows-Installer in a sandbox and examined the extracted-files (also the temp-files) and did a tiny check of the used certificates and the signed files and found some strange issues with the certficate-structure of this project:

## Strange installation of ROOT-CERTIFICATE in ROOT-store

![cert-issues](https://github.com/rustdesk/rustdesk/assets/83558069/a0517017-e01b-4202-bcd2-799a23658b3c)

The certificate (fingerprint: d1dbb672d5a500b9809689caea1ce49e799767f0), that will be used to install and sign the driver under Windows ("rustdeskidddriver.cat") seems to be strange. It is a **self-signed test-certificate by an anonymous id: "WDKTestCert admin,133225435702113567" that will not use a trusted ROOT-CA and only uses SHA-1 with RSA 2048 (weak) with a timed validity of !! 10 Years !!, much too long for this weak algo !**

```
 Fingerprint: d1dbb672d5a500b9809689caea1ce49e799767f0
 CN = WDKTestCert admin,133225435702113567
```

### The fact, that this certificate will be installed in the SYSTEM ROOT-CERTSORE is probably associated with a high risk ...

 Also the certificate, that signed the installed driver-dll **"RustDeskIddDriver.dll"** is not the same as the certifcate, that created / signed the catalog-file above:

```
 Fingerprint: b27892e774a8cbe382ba7ddcf4649a723b6532d8
 CN = Zhou Huabing
 O = Zhou Huabing
 S = Central Singapore
 C = SG
```

This certificate was issued by "Sectigo Public Code Signing Root R46" ...

### How to Reproduce

Check used certfile "RustDeskIddDriver.cer" of the installer and certificate of signed dll "RustDeskIddDriver.dll".

### Expected Behavior

Usage of valid software-signatures, issued by a trusted ROOT-CA with "secure" algorithms for driver-installation and code-signing.

### Operating system(s) on local side and remote side

Windows 10

### RustDesk Version(s) on local side and remote side

1.2.3

### "RustDeskIddDriver.cer"

![cert-issues](https://github.com/rustdesk/rustdesk/assets/83558069/f79774a0-3afe-43b5-99d6-eb1cf4f8ad30)

### RustDesk_install.bat

![cert-issues-02](https://github.com/rustdesk/rustdesk/assets/83558069/e7aa6a0c-5276-4cc7-bd1f-90af6c60368d)

### "RustDeskIddDriver.dll"

![cert-issues-03](https://github.com/rustdesk/rustdesk/assets/83558069/b5af770e-8b7b-4a74-bef5-2e4fd889daa8)

### "rustdeskidddriver.cat"

![cert-issues-04](https://github.com/rustdesk/rustdesk/assets/83558069/f91cb185-6355-4a1b-81e8-a04153ffe6e9)

**Comment by ForDefault** [comment]
I actually ended up here from searching online, after, seeing this weird Cert but mine was in Chinese. 
Same CN=WDKTestCert admin,133225435702113567
Thats all I can contribute. Except, maybe a bump to this lol

**Reply by rustdesk** [reply]
The certificate required to sign the video driver is a Windows WDK certificate and not a Chinese one

**Reply by nOrphf** [reply]
Hi

Glad to read that you have retrieved the EV certificate.
But I also came to this tread after discovering a "Chinese certificate"
![image](https://github.com/rustdesk/rustdesk/assets/54571822/6316a888-8e7c-4424-8bdb-1be5f384c5ad)

Is this still RustDesk certificate?

Regards Lars

**Reply by xlionjuan** [reply]
https://github.com/rustdesk/rustdesk/wiki/Remove-RustDesk-test-certificate

**Comment by deep-soft** [comment]
The certificate (fingerprint: d1dbb672d5a500b9809689caea1ce49e799767f0) is automatically used to sign the driver
https://github.com/fufesou/RustDeskIddDriver/releases/download/v0.4/RustDeskIddDriver_x64.zip
from https://github.com/fufesou/RustDeskIddDriver
All Windows drivers must be signed, this is mandatory.

The certificate (fingerprint: b27892e774a8cbe382ba7ddcf4649a723b6532d8) was used to sign .dll and .exe from the rustdesk package.
https://github.com/rustdesk/rustdesk

**Reply by x011** [reply]
The certificate purposes option in `Trusted Root Certification Authorities` is checked as `Enabled for All Purposes`. Shouldn't it be only for Code Signing?

**Reply by x011** [reply]
HN discussion: https://news.ycombinator.com/item?id=39256493

**Comment by Staja** [comment]
I'm seeing this discussion spread over the internet like wildfire but, when I checked, the certificate only had an EKU for Code signing and KU for Key Encipherment and Data Encipherment; this is what I would expect to see from a certificate used for the stated purposes. TLS stacks (Eg: SChannel) should be adhering to those KU/EKU's; "Enabled for All Purposes" implies "-that are listed in KU/EKU's".

Regardless though, I think a certificate being pushed to end users should clearly identify itself and not use SHA-1, that should be rectified.

**Reply by Zakkaruu** [reply]
To add to this, you can verify and limit "purposes" by going into the certificate properties. The driver certificate only has Code Signing as an option.

<img width="401" alt="image" src="https://github.com/rustdesk/rustdesk/assets/20016567/c0d643ea-aeda-4f4d-8f5d-4d66aa0bd8b5">

Compare with an all purposes, all applications root certificate:

<img width="397" alt="image" src="https://github.com/rustdesk/rustdesk/assets/20016567/a48e7d97-356a-4449-8277-15c1e6954e79">

For MITM, Key Usage would need to have Certificate Signing. The RustDesk cert does not.

<img width="386" alt="image" src="https://github.com/rustdesk/rustdesk/assets/20016567/34cfa73f-f6e5-4821-b644-303dc4bb02e8">

So long as the properties are respected as Staja said. _(I tested with my own MITM/decryption cert, and they are respected in browsers--didn't test anything else)_

I am also inclined to believe that the certificate just needs some fine tuning for transparency and increased security.

**Comment by rustdesk** [comment]
Ok, we will remove this idd driver since it brings so big trouble. Some functions (virtual display, privacy mode etc) will be lost.

**Reply by Hydro8** [reply]
Maybe just quickly explain :
- why this certificat is in root store. 
- why this certificat have a validity of 10 years and only use SHA-1. 
- why this certificat is different from the DLL one.

**Reply by rustdesk** [reply]
Frankly, as for 1,2, we do not know either, we are not expert in this field, maybe because it is test cert.
As for 3, dll is OV cert, but the driversigning requires EV cert, we do not have EV cert, so we use test cert as a workaround.
Someone in HN does not believe we do not have EV, but we really have not. :(
But now I think it is time for us to get one.

**Reply by Hydro8** [reply]
The main problem is to be sur that we can use rustdesk without backdoor or some problems with privacy, especially in business area.

I understand that you can't know everything. I think this is where open source community is really powerful.
You're not alone ;)

**Reply by xlionjuan** [reply]
Hope it could come back when 1.2.4 released😢

**Reply by x011** [reply]
If you guys are serious about RustDesk, it's time to buy an EV certificate, it costs $270 per Year.

**Reply by Hydro8** [reply]
Do you know the list of features that will be lost ?

**Reply by xlionjuan** [reply]
@Hydro8 virtual display and privacy mode

**Reply by danielmarschall** [reply]
When RustDesk is installed, there is a checkbox asking if the certificate should be installed or not. Wouldn't it be a solution to enable the features if the user chooses to install the cert, and disable the features if the user denies it?

**Reply by danielmarschall** [reply]
If the Virtual Display feature is removed in the next version, would that mean that I can no longer establish an unattended connection to a terminal server that has no attached monitor and/or active user session?

**Reply by florianbeisel** [reply]
> Frankly, as for 1,2, we do not know either, we are not expert in this field, maybe because it is test cert.

>* why this certificat is in root store.

Because its explicitly installed there by `install_cert_runas_admin.bat` in the [RustDeskIddDriver](https://github.com/fufesou/RustDeskIddDriver).

```bash
.\CertMgr.exe /add RustDeskIddDriver.cer /s /r localMachine root
```

**Comment by rustdesk** [comment]
> When RustDesk is installed, there is a checkbox asking if the certificate should be installed or not. Wouldn't it be a solution to enable the features if the user chooses to install the cert, and disable the features if the user denies it?

This is the current behavior, but it can not stop thread like this.

**Reply by danielmarschall** [reply]
@rustdesk Thank you for your quick reply. Can you please tell me if the Virtual Display feature is removed in the next version, would that mean that I can no longer establish an unattended connection to a terminal server that has no attached monitor and/or active user session? Or is the Virtual Display Driver just to make things a bit prettier (resolution, etc.)?

**Comment by joru1407** [comment]
This driver which is only signed with the Test-Certificate also leads to issues with ESET for example. ESET detects the RustDeskIddDriver.dll under %LOCALAPPDATA%\rustdesk\RustDeskIddDriver as Suspicious Object, moves it to the Quarantine and warns the user about it.

**Comment by TGeRi** [comment]
Hi! This bug is still present in a way!
When I install Rustdesk (official build) v1.2.3-1 on a fresh PC (test PC is Win11 23H2 for me) the cert IS NOT installed, but a new CERTIFICATE STORE (with the chinese text as the name of this "certificate folder") IS created in Windows. It is empty, but Rustdesk install creates it nontheless.

![image](https://github.com/rustdesk/rustdesk/assets/16209303/16f0a75e-a747-422c-bf44-41d901a87e1f)




It could be concerning for users as Google translate is this for that chinese text:

![image](https://github.com/rustdesk/rustdesk/assets/16209303/deb7d588-2b61-4467-b462-feb64c84c462)
:)

## Correct
This is not chinese text, it is garble code.

**Reply by fufesou** [reply]
This is so strange.

I checked the code of 1.2.3-1 again and it will not try install the cert.

Unless someone has administrator rights and manually executes the command, this is equivalent to using certmgr.exe.

https://github.com/rustdesk/rustdesk/blob/a71ec3e0e740c69c188641d4818d606201bbe1b4/src/core_main.rs#L205

Did you only install 1.2.3-1 and no other versions of RustDesk on your fresh PC?
1.2.3-1 does not have a certificate file, and install_cert will return at the entrance.

https://github.com/rustdesk/rustdesk/blob/a71ec3e0e740c69c188641d4818d606201bbe1b4/src/platform/windows.rs#L1923


Did you perform any other operations on your new computer?
I think this is more likely due to other reasons.

**Reply by danielmarschall** [reply]
Actually, I have this too. But I don't know when it was created. Could have been before 2.3.1-1. I have two roots, one with a certificate in it, and one without. I have now deleted the certificate, uninstalled rustdesk, and installed nightly, and it is NOT created again.

![image](https://github.com/rustdesk/rustdesk/assets/28412477/256b745d-aefc-4532-a949-363f3847cf14)

Unfortunately, I cannot delete the branches! They are there, even if no certificate is inside. How can I delete them?

EDIT: Additional note: Actually, I had the WDKTest certificate in this Chinese Characters Certificate Store, as well as in the "Trusted Publishers Certificate Store".

**Reply by fufesou** [reply]
Sorry for causing trouble to you guys.

You can delete strange branches by deleting the corresponding entries in regedit.

1. Open "certmgr.msc"
2. Open "regedit", delete the strange items. 
3. Close and open "certmgr.msc".

If there're stille some strange items, you may not delete all items in "regedit".
Refresh "regedit" and delete again.

https://github.com/rustdesk/rustdesk/assets/13586388/8bd868c9-381d-4402-ae42-0abc6c54e895


**NOTE**

There should be more than one place, you may need to find "Ctrl + F" to search the registry.
```
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\SystemCertificates
Computer\HKEY_CURRENT_USER\Software\Microsoft\SystemCertificates
Computer\HKEY_CURRENT_USER\Software\Microsoft\SystemCertificates\UserDS
```

**Reply by fufesou** [reply]
@TGeRi Confirmed. My PC creates this strange "certificate folder" too, after installing 1.2.3-1.

I'll try find out what causes this strange behavior.

**Reply by fufesou** [reply]
This code causes the strange cert directory.

https://github.com/rustdesk/rustdesk/blob/a71ec3e0e740c69c188641d4818d606201bbe1b4/src/platform/windows.rs#L1982

The corresponding C++ code is
```C++
#pragma comment(lib, "crypt32.lib")

#include <Windows.h>
#include <iostream>

int main(int argc, char* argv)
{
	HCERTSTORE handle = CertOpenSystemStoreW(NULL, (LPCWSTR)"ROOT");
	return 0;
}
```

1.2.3-1 will not install the cert. But it will try to remove the old test certs.

It tries to read the certs in `ROOT` , then find the issuer of `"CN=\"WDKTestCert admin,133225435702113567\"\0"`. Remove all it's related certs.
It will found no certs, and do nothing. But the directory is created and left there.

I'll try the code https://github.com/rustdesk/rustdesk/discussions/6444#discussioncomment-9014382 to remove the strange directories.

**Comment by rustdesk** [comment, marked as answer]
# Fixed in last year by removing our virtual display driver and cert.

For the time being we use usbmmidd_v2 which is signed and verified by $MS

We might switch to another better choice, but not decide yet.
https://github.com/rustdesk/rustdesk/discussions/10601
